pkg/virt-controller/watch/node.go

```diff
@@ -165,65 +163,194 @@ func (c *NodeController) execute(key string) error {
-			_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, []byte(fmt.Sprintf("[%s, %s]", phasePatch, reasonPatch)))
-			if err != nil {
-				errs = append(errs, fmt.Sprintf("failed to move vmi %s in namespace %s to final state: %v", vmi.Name, vmi.Namespace, err))
-				logger.Reason(err).Errorf("Failed to move vmi %s in namespace %s to final state", vmi.Name, vmi.Namespace)
-			}
-		}
+	if len(errs) > 0 {
+		return fmt.Errorf("%v", strings.Join(errs, "; "))
+	}
 
-		if len(errs) > 0 {
-			return fmt.Errorf("%v", strings.Join(errs, "; "))
-		}
+	return nil
+}
+
+func (c *NodeController) createAndApplyFailedVMINodeUnresponsivePatch(vmi *virtv1.VirtualMachineInstance, logger *log.FilteredLogger) error {
+	c.recorder.Event(vmi, v1.EventTypeNormal, NodeUnresponsiveReason, fmt.Sprintf("virt-handler on node %s is not responsive, marking VMI as failed", vmi.Status.NodeName))
+	logger.V(2).Infof("Moving vmi %s in namespace %s on unresponsive node to failed state", vmi.Name, vmi.Namespace)
+
+	patch := generateFailedVMIPatch(vmi.Status.Reason)
+	_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, patch)
+	if err != nil {
+		logger.Reason(err).Errorf("Failed to move vmi %s in namespace %s to final state", vmi.Name, vmi.Namespace)
+		return err
 	}
-	if nodeExists {
-		c.Queue.AddAfter(key, c.recheckInterval)
+
+	return nil
+}
+
+func generateFailedVMIPatch(reason string) []byte {
+	phasePatch := fmt.Sprintf(`{ "op": "replace", "path": "/status/phase", "value": "%s" }`, virtv1.Failed)
+	operation := "add"
+	if reason != "" {
+		operation = "replace"
+	}
+	reasonPatch := fmt.Sprintf(`{ "op": "%s", "path": "/status/reason", "value": "%s" }`, operation, NodeUnresponsiveReason)
+	return []byte(fmt.Sprintf("[%s, %s]", phasePatch, reasonPatch))
+}
+
+func (c *NodeController) requeueIfExists(key string, node *v1.Node) {
+	if node == nil {
+		return
+	}
+
+	c.Queue.AddAfter(key, c.recheckInterval)
+}
+
+func (c *NodeController) markNodeAsUnresponsive(node *v1.Node, logger *log.FilteredLogger) error {
+	c.recorder.Event(node, v1.EventTypeNormal, NodeUnresponsiveReason, "virt-handler is not responsive, marking node as unresponsive")
+	logger.V(4).Infof("Marking node %s as unresponsive", node.Name)
+
+	data := []byte(fmt.Sprintf(`{"metadata": { "labels": {"%s": "false"}}}`, virtv1.NodeSchedulable))
+	_, err := c.clientset.CoreV1().Nodes().Patch(context.Background(), node.Name, types.StrategicMergePatchType, data, metav1.PatchOptions{})
+	if err != nil {
+		logger.Reason(err).Error("Failed to mark node as unschedulable")
+		return fmt.Errorf("failed to mark node %s as unschedulable: %v", node.Name, err)
 	}
+
 	return nil
 }
 
+func (c *NodeController) createEventIfNodeHasOrphanedVMIs(node *v1.Node, vmis []*virtv1.VirtualMachineInstance) error {
+	// node is not running any vmis so we don't need to check anything else
+	if len(vmis) == 0 || node == nil {
+		return nil
+	}
```

Comment:
one thing to possibly be aware of here. I believe finalized VMIs might be in this list. That's probably the case if we're simply getting all VMIs that have a label selector where the kubevirt.io/nodeName==nodename.

So, there are cases where VMIs might be dead but not garbage collected for whatever reason. Then this event will get fired if virt-handler ever goes down on that node intentionally.